packages/angular_devkit/schematics/src/tree/host-tree.ts

```diff
@@ -299,6 +301,19 @@ export class HostTree implements Tree {
     return data.toString();
   }
 
+  readJSON(path: string): JsonValue | undefined {
+    const content = this.readText(path);
+    if (content === null) {
+      // undefined is returned for the not found case because null is a valid JSON value
+      return undefined;
+    }
```

Comment:
This is really awkward, but I think your approach is probably the best one.

Did you consider changing `readText()` and `read()` to return `undefined` as well? I'm guessing that's more complicated because `read()` is already public API and that would be a breaking change. Just wondering if it would be worth making that change.

Reply:
I agree.  I don't really like this either.
I have an alternative form of readText/readJson that throws instead which i think i'll bring into this PR.  From reworking existing schematics i think this better represents the usages of `read` even though the semantics will now differ from `read`.  Code similar to the following is repeated extensively throughout schematics (which could all be reduced to a single `readText` call):
```
  const text = host.read(modulePath);
  if (text === null) {
    throw new SchematicsException(`File ${modulePath} does not exist.`);
  }
  const sourceText = text.toString('utf-8');
```

Reply:
+1 to the error. I wish we could use some kind of `Result<string, SchematicsException>` here, but given the current state of JS, I think the error is the best way to go here.